### DISCUSSION TITLE
Floats don't round as the layman would expect

### DISCUSSION THREAD

**Discussion by matterhorn103** [mdickinson/rounders#28] [General]
Hi! I am very much aware that you are an actual CPython developer, so I hope I do not come across too insolent here. I fully expect you to just tell me all the ways in which I am wrong!

This seems like a great package that would solve all my (and many other people's) issues, and prevent me having to write complicated code using Decimal, the sigfig package, and overriding the rounding behaviour of Decimal just in order to get sensible values written to csv files at the end of scientific scripts. The code often looks like this (working with pandas dataframes):

```scans.loc[v, "j_red_f"] = float((Decimal(str(df.iloc[0]["Peak height"])) / Decimal(electrode_area_in_cm2).quantize(Decimal("0.00001"), rounding=ROUND_UP)))```

when with your package it could look like:

```scans.loc[v, "j_red_f"] = round(df.iloc[0]["Peak height"] / electrode_area_in_cm2, 5)```

And that's without even considering that actually I'd like to be using significant figures rather than decimal places for that particular line.

Unfortunately though the rounding method used for floats still gives unintuitive output to the average scientific user of Python (imo) i.e. someone unacquainted with the binary issue.

The outcome of rounding floats using rounders.round() is presumably the way it is because instead of the printed representation of e.g. 0.15 being used, the actual stored value of, in this case, 0.149999999999999994448884876874217297881841... is used, meaning that here for example the `TIES_TO_X` rounding modes always give 0.1, even when 0.2 would be expected.

This can be seen using:
```Python 3.10.12 (main, Jun 30 2023, 17:58:40) [GCC] on linux
>>> from decimal import Decimal
>>> from rounders import round, TIES_TO_AWAY
>>> for x in range(19):
...         y = round((x / 10) + 0.05, 2)
...         print(y, "--->", round(y, 1, mode=TIES_TO_AWAY), "while input as Decimal:", Decimal(y))
```
```
0.05 ---> 0.1 while input as Decimal: 0.05000000000000000277555756156289135105907917022705078125
0.15 ---> 0.1 while input as Decimal: 0.1499999999999999944488848768742172978818416595458984375
0.25 ---> 0.3 while input as Decimal: 0.25
0.35 ---> 0.3 while input as Decimal: 0.34999999999999997779553950749686919152736663818359375
0.45 ---> 0.5 while input as Decimal: 0.450000000000000011102230246251565404236316680908203125
0.55 ---> 0.6 while input as Decimal: 0.5500000000000000444089209850062616169452667236328125
0.65 ---> 0.7 while input as Decimal: 0.65000000000000002220446049250313080847263336181640625
0.75 ---> 0.8 while input as Decimal: 0.75
0.85 ---> 0.8 while input as Decimal: 0.84999999999999997779553950749686919152736663818359375
...
```

As noted in the docs on floats, Python is already (since 3.1) clever enough to simplify the output of a float, via the `__repr__` method, such that `print(float(0.15))` prints 0.15 rather than the 17-place decimal. Related to this is that `str(float(0.15))` gives "0.15". This means that while `Decimal(float(0.15))` gives` Decimal("0.14999999...")`, `Decimal(str(float(0.15)))` gives `Decimal("0.15")`. It seems like it would thus be relatively easy to implement it so that the float is treated as the representation given to the user, rather than the stored value.

Unfortunately, my Python skills aren't good enough for me to really understand exactly what's going on in the underlying code, but it seems like the first step is to turn the input number into an IntermediateForm object that is defined purely in terms of integers. Can't these integers just be obtained from the string representation of the float?

Indeed I implemented a working example myself with my own incredibly limited abilities, which has the desired behaviour for floats:
```Python
from decimal import Decimal

def round(x, places=0):
    y = str(x)
    pre_point = y.split(".")[0]
    post_point = y.split(".")[1]
    if places == 0:
        if post[0] in [str(n) for n in range(0, 5)]:
            output = int(pre_point)
        elif post[0] in [str(n) for n in range(5, 10)]:
            output = int(pre_point) + 1
    else:
        post_short = int(post_point[:places])
        next_digit = post_point[places]
        if next_digit in [str(n) for n in range(0, 5)]:
            output = float(pre_point + "." + str(post_short))
        elif next_digit in [str(n) for n in range(5, 10)]:
            output = float(pre_point "." + str(post_short + 1))
    return output
```

Obviously, as written the above would fail in many many situations. But why wouldn't an approach along these lines work? What is the harm in treating floats this way?

I actually found your package from the discussion [here](https://discuss.python.org/t/rounding-to-significant-figures-feature-request-for-math-library/16395/87?page=3). I think that, given your package has everything else a user would want from a rounding package (rounding to significant figures, choice of rounding method, works with floats and integers and Decimal and Fraction) making the behaviour of floats beginner-friendly would make it the perfect antidote to a great many people's headaches, no? As a non-expert Python user, you care about the printed value of a float, not the stored value.

If you felt like this abuses floats too much and that rounders.round() should give the same results for floats as built-in round(), would it be possible for there to be a configurable option? So that the package can be loaded with "float_as_repr" or similar to instruct round() to use the string representation?

I get that floats are intrinsically weird, should be accepted for what they are, and one should use Decimal instead for high precision stuff (as many people in that thread said repeatedly), but unfortunately some common scientific packages e.g. numpy/scipy/pandas do not work well with Decimal objects, making constant conversion between float and Decimal necessary. Your package very nearly solves my problem for me!

**Comment by ChrisBarker-NOAA** [comment]
I'm not the developer, nor an FP expert, but I am a scientist, and was part of the referenced discussion.

In short -- what`rounder` does is correct, and should not be changed. It works with floating point numbers, and yes, they are "intrinsically weird" -- though not really -- what they are is binary, and most of us normal folks are continually confused by that :-)

One of the key points (and it came up in the discussion above, is that in real code, `round()` and the like are usually applied to the results of computation, which have some limited precision in the result -- so you may expect to get exactly say, 0.15, when but you really get 0.1499999999999 or 0.150000000000001 -- and chances are, the "error" is in more than one digit at the end. So in fact, if matters little (or none at all) whether that rounds to .1 or .2 --  if it does matter, you need more precision in your calculation!

So on to problem -- and I'm thinking about this, and commenting, because I face the same issues, and am looking for solutions, too.

First: 
> one should use Decimal instead for high precision stuff

This is a (I think dangerous) misconception -- Decimal is not more precise than floats -- Decimal is also floating point, and has the similar limitations. It does have two advantages:

1)  Perception - we all expect "1/10" to be exact, but are not surprised that "1/3" is not -- but float and Decimal can both only exactly capture a finite level of precision.

2) Decimal supports variable precision, so if you do need more precision then you can get that. That can be very useful, but  most of us don't need that -- particularly if you're going to round later anyway :-)

> Decimal(float(0.15)) gives Decimal("0.14999999..."), Decimal(str(float(0.15))) gives Decimal("0.15"). It seems like it would thus be relatively easy to implement it so that the float is treated as the representation given to the user, rather than the stored value.

Well, no -- because you don't know where the float came from -- it *may* have been inputted by the user as a nice round decimal number, but it may not have -- what str() does is round a little tiny bit to make it nicely human readable, but it is actually rounding a bit -- there is no way to know that it's intended to be exactly 0.15, or if it is the result of a calculation that is, ifact, a teeny tiny bit less than 0.15

>  So that the package can be loaded with "float_as_repr" or similar to instruct round() to use the string representation?

hmm -- that's an interesting idea -- but I think you may be disappointed -- first,what repr is doing is very much decimal-based slight rounding -- if there is a simple decimal representation that is less than an epsilon asway from the actual value,then it round it -- which means i t's really a job for Decimal, potentially, not rounder.

However, you may be missing something -- we were told in primary school that we should round up to break the tie: e.g. 0.15 rounds to 0.2 -- however, that would introduce a slight bias in results if it were used extensively in computation, which is why the defulat round does, I think "ties to even", e.g.:

```
In [69]: round(1.5)
Out[69]: 2

In [70]: round(2.5)
Out[70]: 2
```
You know that, which is why you used TIES_TO_AWAY in your example.

But TIES_TO_AWAY is still inherently based on Decimal.

Back to the problem at hand:

I'm trying to understand the context of your problem. To the precision of the computation, rounders is doing the right thing (which probably shouldn't be TIES_TO_AWAY, by the way, but that's your choice)

So I suspect that the problem is that somewhere you are presenting a user, but the iriginal and rounded number:

x = 0.15, rounded to 0.1 # huh????

or you are taking input from a user (either directly, or via a text file or some such, and then presenting the rounded value back.

For case one -- hmm ... I've never done that.

For case two -- what you really want to do is handle it in I/O -- perhaps with Decimal.

As for your example:

first -- clearly define the "desired behaviour" -- i'm not entirely sure what that is -- you may be focusing a bit too much on the simple examples.

second -- if you are going to use Decimal, you should be able to do the:

`x = Decimal(str(x))` trick, and then round in `Decimal`, using the `decimal.ROUND_HALF_UP` mode in the context.

(exercise left for the reader -- I can't get it to work right now :-( )

However, in the end, it seems it would be nice to expose the nifty "round to nearest decimal" function that the __repr__ uses -- maybe in the Decimal module, where it could optionally be used when converting floats to Decimal.

-CHB

**Comment by mdickinson** [comment]
@matterhorn103 @ChrisBarker-NOAA Thanks for the issue and discussion! I'm not planning to change the behaviour of the library here, since rounding based on the value of the floating-point number really _is_ what I want here, so I've migrated the issue to a discussion.

If you want a do-what-I-mean style round, the `Decimal(str(x))` trick that @ChrisBarker-NOAA suggests is the way to go. But note that that's not going to save you from other surprises later on - e.g., even if `round` went via the decimal `str`, you'd still get surprised by the result of `round_ties_to_away(0.15 + 0.3)`, for example. The only way to remove the surprises entirely is to work in decimal all the way through.

**Reply by matterhorn103** [reply]
> e.g., even if round went via the decimal str, you'd still get surprised by the result of round_ties_to_away(0.15 + 0.3), for example. The only way to remove the surprises entirely is to work in decimal all the way through.

I didn't even know that this behaves like this, thanks for the heads up.

And that's fair enough, it's your package. I guess I just lament that working with decimal floating point as opposed to binary is so syntactically tedious and that numpy/pandas don't treat `float` and `Decimal` equally well.

**Reply by mdickinson** [reply]
Yes - if I had infinite time and energy, fixed-precision `decimal64` and `decimal128` dtypes in NumPy would be very nice to have. Decimal literals in Python is another topic that's been brought up a good few times; there's no fundamental reason why it can't happen, but it would need someone with (a _lot_ of) time and energy to push a proposal through to completion.

**Reply by matterhorn103** [reply]
PyArrow has `decimal128`, and there is experimental implementation of it in Polars despite it being not yet at v1.0, which is quite nice and will make data handling with decimals easier. :)

Doesn't help with calculations, though. It's crazy to me that they are missing in NumPy. Python is so widely used in science and NumPy/SciPy are the essential packages for that community, so I'd have thought it would be one of the *first* things to get implemented. Though I suppose physics and computational sciences managed just fine for decades with floats pre-Python. I guess whenever values are measured by instrumentation – rather than people – you always have a long stream of decimal places past the actual precision, and it's not an issue.

**Reply by ChrisBarker-NOAA** [reply]
> Doesn't help with calculations, though. It's crazy to me that they are missing in NumPy. Python is so widely used in science and NumPy/SciPy are the essential packages for that community, so I'd have thought it would be one of the first things to get implemented.

Well, yes, we re all used to binary FP in computational science. But it's not just that -- that fact is that decimal is not inherently any more accurate than binary. Neither can express all rational numbers (even within a range) exactly. And in actual computation, the results are rarely exact -- the "problem" is almost entirely perception: I typed in 1.5 -- why do I get 1.4999999999 back? Ot I do a computation that I know has an exact decimal result, why don't I get that result? But no one is surprised that 1 / 3 doesn't yield and exact result. Anyway, that's my opinion:  I don't think decimal buys you much in actual practical scientific computation.  So why write a bunch of code to get less efficient computation?

BTW: a few years back there was the intention to refactor numpy to make it easier to define and use custom dtypes -- not sure how far that got, but maybe adding a decimal type would not be that heavy a lift at this point.

**Reply by matterhorn103** [reply]
Sure, I get that, you're absolutely right. For calculations it doesn't make any difference as 1.499999999 and 1.5 are effectively the same. But if I am trying to automate some process of simple arithmetic on decimal numbers, I want to obtain the same result as if I'd done it by hand.

To illustrate my point, in one specific use case of mine, I have datasets with a potential *E*_1 and a potential *E*_2 for each item, and each potential is a decimal with precision to 3 decimal places. There's a lot of these data points, enough that I want to add them to my document via a script, not by hand. I also want to calculate for each item the difference between the two potentials and include it as Δ*E*, but rounded to two decimal places. This would take a while by hand, so why not do it with Python? But the document concerned is a paper for a scientific journal, so the last thing I want is people to browse through the paper or use the data themselves and see what appears to be an obvious error.

That's just one example, there are many I'm sure.

Obviously the answer is to use the Decimal type, and that's fine by me, it's just that it's not supported for everything. I guess I just wish that decimal floating point was an equal citizen in Python and as easy to use and widely/deeply implemented as binary.

Please don't take the spirit of my comments the wrong way, by the way, I'm not seeking out arguments here, just enjoying the discussion.

**Reply by ChrisBarker-NOAA** [reply]
> Sure, I get that, you're absolutely right. For calculations it doesn't make any difference as 1.499999999 and 1.5 are effectively the same.

right -- I was addressing the "why wasn't a decimal type been built into numpy long ago?"

> But if I am trying to automate some process of simple arithmetic on decimal numbers, I want to obtain the same result as if I'd done it by hand.

Ahh -- well then, Decimal is clearly the right tool for that. So the point here is that it would be nice for Decimal to be easier to use -- sure. And, to your original point, a "round as I learned in primary school" method would be handy. And by the way, not hard to write for the Decimal type, anyway :-)

**Comment by matterhorn103** [comment]
@ChrisBarker-NOAA Oh I'm well aware of the bias inherent to always rounding up, and that there are less biased options available. But in some branches of science, chemistry being one, numbers are universally expected to be rounded up. So it's nice to have the options in the package, then everyone is happy.

Sure, when doing calculations that result in long floats then the question is moot, as the float probably has far more precision than the number you are rounding to. The problem really applies in the situation that only simple arithmetic with fairly imprecise numbers is being done.

When writing a module or a function that does something it is actually very hard in Python's current situation to provide that the user expects. An example being if I am making a module that deals with electrochemistry. The user can instantiate `Potential` objects or `ReferenceElectrode` objects with known potentials. These potentials are specified and may have been measured to 2 or 3 s.f. Say the user defines two objects, one a reference electrode with potential +0.715 V and the other a measured potential with less precision, say -1.12 V. Finding the measured potential with respect to the reference potential would give -1.835 V, but this is of course higher precision than is legitimate given the low precision of the measurement, so the user will want this value to be rounded to 2 s.f. before writing it to e.g. a LaTeX document or report. It is very hard to do this in Python in a way that will result in -1.84 V, which is what a chemist would write if rounding manually.

**Comment by ChrisBarker-NOAA** [comment]
@mdickinson: thinking on this more, I think @matterhorn103 has a really good point:

There is an inherent conflict between the binary implementation of floats with the fact that we use decimal notation for human-readable IO.

Python took a small step to address this with the choice to have the repr() of floats round to a simple decimal form when the value is "close enough" to a simple decimal form -- and this has been a really nice thing for usability.

But maybe we can take it a step farther, by doing the same thing in other contexts in which we are working with floats in the decimal space. I can think of two of these:

1) converting a `float` to a `Decimal` -- if, e.g. float("0.15") is close enough to the actual value of 0.15 to get diisplayed as such, then maybe (optionally?) it should convert to a `Decimal` with that exact value. In short, so that:

`Decimal(a_float) == Decimal(repr(a_float))`

This is complicated by the fact that Decimal, by default, has higher precision than float. But IIUC, repr(float) does in fact produce a number that is as close to the actual value as float precision will hold (e.g. `float(repr(a_float)) == a_float`), so you woulnd't be losing any actual precision, you'd just get a similarly precise value that is a "nice" Decimal.

Now that I think about it, this might mean that a Decimal literal would not be necessary -- or a lot less necessary.

2) To the OP's point -- for rounding to decimal digits, the almost-decimal conversions could also be used -- to get what might even be more accurate decimal rounding.
